pkg/scheduler/framework/plugins/volumebinding/volume_binding.go

```diff
@@ -87,9 +109,137 @@ func (pl *VolumeBinding) Filter(ctx context.Context, cs *framework.CycleState, p
+// the PV controller has completely finished the binding operation.
+//
+// If binding errors, times out or gets undone, then an error will be returned to
+// retry scheduling.
+func (pl *VolumeBinding) PreBind(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) *framework.Status {
+	state, err := cs.Read(allBoundStateKey)
+	if err != nil {
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	allBound, ok := state.(allBoundStateData)
+	if !ok {
+		return framework.NewStatus(framework.Error, "unable to convert state into allBoundStateData")
+	}
+	if allBound {
+		// no need to bind volumes
+		return nil
+	}
+	klog.V(5).Infof("Trying to bind volumes for pod \"%v/%v\"", pod.Namespace, pod.Name)
+	err = pl.binder.BindPodVolumes(pod)
+	if err != nil {
+		klog.V(1).Infof("Failed to bind volumes for pod \"%v/%v\": %v", pod.Namespace, pod.Name, err)
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	klog.V(5).Infof("Success binding volumes for pod \"%v/%v\"", pod.Namespace, pod.Name)
+	return nil
+}
+
+// Unreserve is called by the scheduling framework when a reserved pod was
+// rejected in a later phase.
+// TODO(cofyc) Revert assumed PV/PVC cache, see http://issues.k8s.io/82934#issuecomment-538269188
+func (pl *VolumeBinding) Unreserve(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) {
+	pl.binder.DeletePodBindings(pod)
+	return
+}
+
+// PostBind is called after a pod is successfully bound.
+func (pl *VolumeBinding) PostBind(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) {
+	pl.binder.DeletePodBindings(pod)
+	return
+}
+
+// Args holds the args that are used to configure the plugin.
+type Args struct {
+	BindTimeoutSeconds *int64 `json:"bindTimeoutSeconds,omitempty"`
+}
+
+// New initializes a new plugin and returns it.
+func New(plArgs runtime.Object, fh framework.FrameworkHandle) (framework.Plugin, error) {
+	args := &Args{}
+	if err := framework.DecodeInto(plArgs, args); err != nil {
+		return nil, err
+	}
+	if err := validateArgs(args); err != nil {
+		return nil, err
+	}
+	var bindTimeoutSeconds int64
+	if args.BindTimeoutSeconds == nil {
+		bindTimeoutSeconds = DefaultBindTimeoutSeconds
+	} else {
+		bindTimeoutSeconds = *args.BindTimeoutSeconds
+	}
+	nodeInformer := fh.SharedInformerFactory().Core().V1().Nodes()
+	pvcInformer := fh.SharedInformerFactory().Core().V1().PersistentVolumeClaims()
+	pvInformer := fh.SharedInformerFactory().Core().V1().PersistentVolumes()
+	storageClassInformer := fh.SharedInformerFactory().Storage().V1().StorageClasses()
+	csiNodeInformer := fh.SharedInformerFactory().Storage().V1().CSINodes()
+	binder := scheduling.NewVolumeBinder(fh.ClientSet(), nodeInformer, csiNodeInformer, pvcInformer, pvInformer, storageClassInformer, time.Duration(bindTimeoutSeconds)*time.Second)
+	// In Filter phrase, pod binding cache is created for the pod and used in
```

Comment:
make this a comment for VolumeBinding struct

Reply:
updated